cas_client/src/remote_client.rs

```diff
@@ -300,45 +302,103 @@ impl RemoteClient {
         fetch_info: Arc<HashMap<HexMerkleHash, Vec<CASReconstructionFetchInfo>>>,
         offset_into_first_range: u64,
         byte_range: Option<FileRange>,
-        writer: &mut Box<dyn Write + Send>,
+        path: &Path,
         progress_updater: Option<Arc<dyn ProgressUpdater>>,
     ) -> Result<u64> {
+        let term_lengths = terms.iter().map(|x| x.unpacked_length as u64).collect::<Vec<_>>();
+        let file_range_per_term = &mut vec![FileRange::default(); terms.len()];
+        for item in file_range_per_term.iter_mut() {
+            item.start = 0;
+            item.end = 0;
+        }
+        if let Some(byte_range) = &byte_range {
+            let mut current_offset = 0;
+            for (idx, (term_range, &length)) in file_range_per_term.iter_mut().zip(term_lengths.iter()).enumerate() {
+                // FIXME I don't understand why the current logic does not take the start range into account
```

Comment:
IIUC, the start range is unneeded since the terms will be received from CAS pruned to the range already, meaning that we don't know how many terms / bytes were skipped to get to the first term in `byte_range.start` is the offset globally across the complete file, however, the terms we are given may not start with the first term in `terms`. This is why we have the `offset_into_first_range` instead of relying on `byte_range.start`.

Reply:
Super clear! Thanks for the explanation 🙏 !